### ISSUE TITLE
Missing "Payer" object in Agreement

### ISSUE THREAD

**Issue by rovenko** [paypal/PayPal-PHP-SDK#155]
The "Agreement" object should return the "payer" data (as per documentation)
https://developer.paypal.com/webapps/developer/docs/api/#agreement-object

However this data is not coming back neither for Billing Agreements created with Credit Cards nor for the Billing Agreements created with PayPal (and it should based on the documentation)
https://developer.paypal.com/webapps/developer/docs/api/#execute-an-agreement

This is a really big issue as there's no way of getting user's name/email/address via API when executing Billing Agreement via PayPal (unless these are collected separately, which defeats the purpose of PayPal checkout)

**Comment by kenjohnson** [comment]
I notice plan description is also missing from the agreement object.  
 $agreement->getPlan()->getDescription returns NULL. 
 I am using release PayPal-PHP-SDK-0.16.1

**Comment by RedLikeTheColor** [comment]
I just ran into this issue myself, and i see it's been open a while. Is there any news?

**Comment by gregwym** [comment]
"the API team is adding this feature in their next release." Has the release been released?

**Comment by jziaja** [comment]
@RedLikeTheColor, @gregwym, sorry for the delay in responding.

> the API team is adding this feature in their next release.

That was the feedback we received from the team back when we originally reported this issue.  Not only has the API team not added support for this feature as of yet, they recently made the decision to  push support for it back.  We're in the process of following up with them to see if we can get better guidance on when this support will be added.  As soon as I have more information, I'll be sure to share it here.

For what it's worth, the SDK team is just as frustrated that this hasn't been resolved yet.

**Comment by gregwym** [comment]
Hi @jziaja, thank you for the update. I wonder is there a release notes on each update of the Reat API? So we know what they have been working on and how frequent they are releasing updates for the API. It would be really helpful if the API team can communicate directly with the API consumers (instead of going through an medium, the SDK team).

**Comment by jziaja** [comment]
@gregwym, I totally agree with you, and you're definitely not the first person to ask for something like this. :)  Unfortunately, there's no place currently where we provide API release notes publicly other than the information that's available on the [PayPal Developer](https://developer.paypal.com/) site.  I'll pass this feedback along internally to see if we can consider adding something like this in the future, though.

**Comment by RedLikeTheColor** [comment]
@jziaja , thanks for the update.

And I feel for you; I've been in that situation before, where I was on a team whose work was waiting on another team. In fact, the other team in my case was waiting on a third party content provider, who in turn was waiting on the US Federal Government, so it was a lot of fun answering questions about why my part of the project wasn't ready.

**Comment by bcallaars** [comment]
178 days have passed on this "api bug" and no progress yet? They couldn't even just change the documentation in the meantime.

**Comment by dessant** [comment]
Actually it's even older than that https://github.com/paypal/PayPal-Python-SDK/issues/57.
It's beyond me why the docs are not updated with a note about this bug.

**Comment by jhalt** [comment]
An update on this issue would be appreciated.

**Comment by jaypatel512** [comment]
Hey @jhalt ! I feel bad to reply here, but there has not been much activity on this bug. It has definitely been planned out, but the tentative date is nowhere close. The reason is, that there are few higher priority API features being planned out and implemented, that needs to be completed before this issue is taken into action. We are trying to see if there is any workaround that could let you guys have that information. We, SDK team are willing to make any such modifications to our SDK to somehow get the Payer information to the subscription team. Hope you understand.

**Comment by jhalt** [comment]
@jaypatel512 We're just going to revert to the old API for now. The update is greatly appreciated, though.

**Comment by bcallaars** [comment]
Yeah, just change the docs and we would be such much happier!

**Comment by jaypatel512** [comment]
- I will keep you updated @jhalt when we have that fixed. Apologies for the confusion, and delay.
- Already started the process @bcallaars Should be up soon.

**Comment by kelonye** [comment]
Hmm, is there then any workaround for associating billing agreements and webhook events to a particular user?

**Comment by kenjohnson** [comment]
kelonye,
You could use IPN. IPN appears to return the
same kind of raw data that it did in x.com.
One of the things it returns is 
recurring_payment_id which is the same as 
agreement_id in REST. When you create your
agreement_id, you could save a lot of user
information with it, then when you get the
agreement_id back in IPN you can use that as 
a key to get at all the user info you saved.
Note, you would have to have a unique agreement_id
for every user, I do that.

**Comment by kelonye** [comment]
@kenjohnson that's super helpful! I guess it's IPN for now.

**Comment by lunmatu101** [comment]
how to get payer info from id plan :-? i still cant get it. so sad : (

**Comment by jasonsavard** [comment]
I thought as an alternative I could use webhook events, but they are not sending any events related to agreements or subscriptions :(  ... they do work for simple payments (non-recurring)

Any status on fixing this bug?

**Comment by jaypatel512** [comment]
hey @jasonsavard ! our team is as anxious on getting this fix made. We are pushing internal teams to expedite this. However, you could help us by questioning and requesting [PayPal Merchant Technical Services](https://www.paypal-techsupport.com/app/ask) with the above questions.

**Comment by samtregar** [comment]
It's amazing that such a critical bug is still unfixed.  It's impossible to use the REST API for subscriptions without this piece.

**Comment by jaypatel512** [comment]
I completely agree with you. We are trying from our end to push for this bug. However, you directly requesting this to [PayPal Merchant Technical Support](https://www.paypal-techsupport.com/app/ask) would definitely help us prove that merchants and developers are indeed blocked by it.

**Comment by samtregar** [comment]
Yup, I've done that a number of times.  I just did it again now.

**Comment by jhalt** [comment]
This issue has been open for nearly a year now, and like @samtregar, I have made multiple requests to PayPal Merchant Technical Support to no avail..

**Comment by bcallaars** [comment]
When it reaches a year, we should send the team a cake for letting a bug
have an anniversary.

**Comment by dessant** [comment]
@bcallaars, it has reached a year, we've missed the anniversary. ;) https://github.com/paypal/PayPal-Python-SDK/issues/57

**Comment by bcallaars** [comment]
Haha oh yes, it's that bad unfortunately.

**Comment by dessant** [comment]
It would be in our best interest to help them find a name for their baby creature. 😋 #NameTheBaby

**Comment by bcallaars** [comment]
How about Careless? #NameTheBaby

**Comment by dessant** [comment]
@bcallaars, it has a mighty sound to it.

My shot is *Payerless Mole*. #NameThePayPalBaby

**Comment by tomwhipple** [comment]
@jhalt @samtregar @dessant I understand your frustration. If you have MTS ticket IDs or emails would you please send them to me `twhipple@paypal.com` so I can investigate?

(Sorry, but I am going to have to clean up the off-topic comments... :frowning:)

**Comment by dessant** [comment]
It would be more productive if we would get a real response, explaining why the docs at least haven't been updated in the past year, and what is generally happening.

Giving blanket responses will not help anyone. I feel after a year of wasting developers' time with this bug, we are right in expecting to get a reply from a human.

**Comment by samtregar** [comment]
@tomwhipple - sent to you.  They told me the same thing they always tell me, they're working on it.

**Comment by engrsandeep** [comment]
Is there any update on this issue?

**Comment by samtregar** [comment]
I'm wondering the same thing, @tomwhipple!  Maybe you could ask someone who works on the team that's not fixing this to come to github and explain what's going on?  Thanks!

**Comment by ercmcd** [comment]
This is definitely causing problems integrating with this and using the rest api for subscription payments.  Any additional information at this time?

**Comment by samtregar** [comment]
Yup, it basically makes it impossible to use the subscriptions API without
also using the classic API at the same time.  This is great fun for me
since it means I need to get my clients to give me credentials for two
completely different APIs.  Thanks PayPal.

**Comment by jeffharrell** [comment]
I recently took over the payments and billing teams at PayPal, so let me help get some attention here. Thanks @tomwhipple for bringing it to my line of sight. 

Unfortunately, I'm can't comment on the history here, but I can say that I want the API teams to have more of a developer feedback loop and that starts here. If I followed the thread correctly, we need to:

- [ ] Update the REST docs to reflect that payer is not *currently* returned with a Billing Agreement
- [ ] Update the API to return `payer`

If that's correct, let me check in and I will provide some updates.

**Comment by samtregar** [comment]
That sounds correct to me @jeffharrell.  You might also add a pointer in the REST docs to the only way t get payer info now - calling get_express_checkout_details in the classic API.  The only way to figure this out now is to find threads like this one online and piece it together.

Exciting!  Maybe this will get fixed someday soon!

**Comment by jeffharrell** [comment]
Ok, cool. I owe an update back to the issue, but it will likely take a day or two.

By the way, @bcallaars if you send a cake I promise to post back pics of the team eating it :smile:

**Comment by samtregar** [comment]
So, how's it going @jeffharrell?

**Comment by KunaalATrehan** [comment]
We are working on the same and fix should be available soon .
Tentatively release date is last week of October,2015

**Comment by jeffharrell** [comment]
@KunaalATrehan beat me to it :) 

I also just realized I need to fix my GH notifications...

**Comment by samtregar** [comment]
@KunaalATrehan @jeffharrell Another week, another ping.  What's up with this bug?

**Comment by kelonye** [comment]
As soon as Stripe sets up shop in my country ...

**Comment by jeffharrell** [comment]
The work is in progress and it was mentioned above that the fix should be live in the last week of October, which is next week. 

@KunaalATrehan do you have clarity into *when* that will be?

**Comment by SaravananThandapani** [comment]
The fix will go live by 29th October and will keep you updated.

**Comment by samtregar** [comment]
So exciting!

**Comment by SaravananThandapani** [comment]
The fix has been pushed to live today. please verify and let us know..

**Comment by jaypatel512** [comment]
Verified that the payer_info is returned back when subscription is executed for paypal subscriptions. Closing this issue.

**Comment by jhalt** [comment]
Thank you guys!

**Comment by samtregar** [comment]
This is awesome!  Finally got around to implementing code for this and it worked perfectly.  Thank you so much!

There's just one problem though, I still have to use the classic API to fully support recurring payments because that's the only way (as far as I know) to refund individual transactions from a recurring billing agreement.  I added an issue with the details here:

https://github.com/paypal/PayPal-Python-SDK/issues/115

Let me know if I filed it in the right place or if reporting it in another way would help you.

Thanks!
